src/cache/disk.rs

```diff
@@ -18,17 +18,59 @@ use futures_03::executor::ThreadPool;
 use lru_disk_cache::Error as LruError;
 use lru_disk_cache::LruDiskCache;
 use std::ffi::OsStr;
+use std::mem;
 use std::path::{Path, PathBuf};
 use std::sync::{Arc, Mutex};
+use std::thread::{self, JoinHandle};
 use std::time::{Duration, Instant};
 
 use crate::errors::*;
 
+enum DiskCacheInit {
+    Initializing(JoinHandle<LruDiskCache>),
+    Joining,
+    Initialized(Arc<Mutex<LruDiskCache>>),
+}
+
+impl DiskCacheInit {
+    fn new(root: &OsStr, max_size: u64) -> Self {
+        let root = root.to_os_string();
+        let worker = thread::spawn(move || {
+            LruDiskCache::new(root, max_size).expect("Couldn't instantiate disk cache!")
+        });
+
+        Self::Initializing(worker)
+    }
+
+    fn opt_lru(&self) -> Option<&Arc<Mutex<LruDiskCache>>> {
+        if let Self::Initialized(lru) = self {
+            Some(lru)
+        } else {
+            None
+        }
+    }
+
+    fn lru(&mut self) -> &Arc<Mutex<LruDiskCache>> {
+        let state = mem::replace(self, DiskCacheInit::Joining);
+        if let Self::Initializing(worker) = state {
+            let lru = worker.join().unwrap();
+            *self = DiskCacheInit::Initialized(Arc::new(Mutex::new(lru)));
+        } else {
+            *self = state;
+        }
+
+        if let Self::Initialized(lru) = self {
+            lru
+        } else {
+            unreachable!("The disk cache should have been initialized")
+        }
+    }
+}
+
 /// A cache that stores entries at local disk paths.
-#[derive(Clone)]
 pub struct DiskCache {
     /// `LruDiskCache` does all the real work here.
-    lru: Arc<Mutex<LruDiskCache>>,
+    lru_init: Mutex<DiskCacheInit>,
```

Comment:
This makes it necessary to take 2 mutexes for any normal access to the disk cache, which is not a huge thing considering disk will be slow anyways, but sounded enough of a burden code-wise that I looked into it more closely, and it turns out it's not necessary to initialize in a background thread at all. All you really need is to not initialize in `DiskCache::new`. IOW, lazy initialization would work just as well.
My patch looks like this:
```diff
diff --git a/src/cache/disk.rs b/src/cache/disk.rs
index 6a49062..2a6285f 100644
--- a/src/cache/disk.rs
+++ b/src/cache/disk.rs
@@ -17,30 +17,47 @@ use crate::util::SpawnExt;
 use futures_03::executor::ThreadPool;
 use lru_disk_cache::Error as LruError;
 use lru_disk_cache::LruDiskCache;
-use std::ffi::OsStr;
+use std::ffi::{OsStr, OsString};
 use std::path::{Path, PathBuf};
 use std::sync::{Arc, Mutex};
 use std::time::{Duration, Instant};
 
 use crate::errors::*;
 
+enum LazyDiskCache {
+    Uninit { root: OsString, max_size: u64 },
+    Init(LruDiskCache),
+}
+
+impl LazyDiskCache {
+    fn get_or_init(&mut self) -> &mut LruDiskCache {
+        match self {
+            LazyDiskCache::Uninit { root, max_size } => {
+                *self = LazyDiskCache::Init(
+                    LruDiskCache::new(&root, *max_size).expect("Couldn't instantiate disk cache!"),
+                );
+                self.get_or_init()
+            }
+            LazyDiskCache::Init(d) => d,
+        }
+    }
+}
+
 /// A cache that stores entries at local disk paths.
 #[derive(Clone)]
 pub struct DiskCache {
     /// `LruDiskCache` does all the real work here.
-    lru: Arc<Mutex<LruDiskCache>>,
+    lru: Arc<Mutex<LazyDiskCache>>,
     /// Thread pool to execute disk I/O
     pool: ThreadPool,
 }
 
 impl DiskCache {
     /// Create a new `DiskCache` rooted at `root`, with `max_size` as the maximum cache size on-disk, in bytes.
-    pub fn new<T: AsRef<OsStr>>(root: &T, max_size: u64, pool: &ThreadPool) -> DiskCache {
+    pub fn new<T: AsRef<OsStr>>(root: T, max_size: u64, pool: &ThreadPool) -> DiskCache {
         DiskCache {
             //TODO: change this function to return a Result
-            lru: Arc::new(Mutex::new(
-                LruDiskCache::new(root, max_size).expect("Couldn't instantiate disk cache!"),
-            )),
+            lru: Arc::new(Mutex::new(LazyDiskCache::Uninit { root: root.as_ref().to_os_string(), max_size })),
             pool: pool.clone(),
         }
     }
@@ -59,7 +76,7 @@ impl Storage for DiskCache {
         let key = key.to_owned();
         Box::new(self.pool.spawn_fn(move || {
             let mut lru = lru.lock().unwrap();
-            let f = match lru.get(&path) {
+            let f = match lru.get_or_init().get(&path) {
                 Ok(f) => f,
                 Err(LruError::FileNotInCache) => {
                     trace!("DiskCache::get({}): FileNotInCache", key);
@@ -85,19 +102,19 @@ impl Storage for DiskCache {
         Box::new(self.pool.spawn_fn(move || {
             let start = Instant::now();
             let v = entry.finish()?;
-            lru.lock().unwrap().insert_bytes(key, &v)?;
+            lru.lock().unwrap().get_or_init().insert_bytes(key, &v)?;
             Ok(start.elapsed())
         }))
     }
 
     fn location(&self) -> String {
-        format!("Local disk: {:?}", self.lru.lock().unwrap().path())
+        format!("Local disk: {:?}", self.lru.lock().unwrap().get_or_init().path())
     }
 
     fn current_size(&self) -> SFuture<Option<u64>> {
-        f_ok(Some(self.lru.lock().unwrap().size()))
+        f_ok(Some(self.lru.lock().unwrap().get_or_init().size()))
     }
     fn max_size(&self) -> SFuture<Option<u64>> {
-        f_ok(Some(self.lru.lock().unwrap().capacity()))
+        f_ok(Some(self.lru.lock().unwrap().get_or_init().capacity()))
     }
 }
```

However, both this patch and your patch suffer from the problem that when errors happen during the disk cache initialization (example: can't create the `SCCACHE_DIR`), then sccache fails to shutdown.

Reply:
This version works better:
```
diff --git a/src/cache/disk.rs b/src/cache/disk.rs
index 6a49062..d8707b6 100644
--- a/src/cache/disk.rs
+++ b/src/cache/disk.rs
@@ -17,30 +17,62 @@ use crate::util::SpawnExt;
 use futures_03::executor::ThreadPool;
 use lru_disk_cache::Error as LruError;
 use lru_disk_cache::LruDiskCache;
-use std::ffi::OsStr;
+use std::ffi::{OsStr, OsString};
 use std::path::{Path, PathBuf};
 use std::sync::{Arc, Mutex};
 use std::time::{Duration, Instant};
 
 use crate::errors::*;
 
+enum LazyDiskCache {
+    Uninit { root: OsString, max_size: u64 },
+    Init(LruDiskCache),
+}
+
+impl LazyDiskCache {
+    fn get_or_init(&mut self) -> Result<&mut LruDiskCache> {
+        match self {
+            LazyDiskCache::Uninit { root, max_size } => {
+                *self = LazyDiskCache::Init(LruDiskCache::new(&root, *max_size)?);
+                self.get_or_init()
+            }
+            LazyDiskCache::Init(d) => Ok(d),
+        }
+    }
+
+    fn get(&mut self) -> Option<&mut LruDiskCache> {
+        match self {
+            LazyDiskCache::Uninit { .. } => None,
+            LazyDiskCache::Init(d) => Some(d),
+        }
+    }
+
+    fn path(&self) -> &Path {
+        match self {
+            LazyDiskCache::Uninit { root, .. } => root.as_ref(),
+            LazyDiskCache::Init(d) => d.path(),
+        }
+    }
+}
+
 /// A cache that stores entries at local disk paths.
 #[derive(Clone)]
 pub struct DiskCache {
     /// `LruDiskCache` does all the real work here.
-    lru: Arc<Mutex<LruDiskCache>>,
+    lru: Arc<Mutex<LazyDiskCache>>,
     /// Thread pool to execute disk I/O
     pool: ThreadPool,
 }
 
 impl DiskCache {
     /// Create a new `DiskCache` rooted at `root`, with `max_size` as the maximum cache size on-disk, in bytes.
-    pub fn new<T: AsRef<OsStr>>(root: &T, max_size: u64, pool: &ThreadPool) -> DiskCache {
+    pub fn new<T: AsRef<OsStr>>(root: T, max_size: u64, pool: &ThreadPool) -> DiskCache {
         DiskCache {
             //TODO: change this function to return a Result
-            lru: Arc::new(Mutex::new(
-                LruDiskCache::new(root, max_size).expect("Couldn't instantiate disk cache!"),
-            )),
+            lru: Arc::new(Mutex::new(LazyDiskCache::Uninit {
+                root: root.as_ref().to_os_string(),
+                max_size,
+            })),
             pool: pool.clone(),
         }
     }
@@ -59,7 +91,7 @@ impl Storage for DiskCache {
         let key = key.to_owned();
         Box::new(self.pool.spawn_fn(move || {
             let mut lru = lru.lock().unwrap();
-            let f = match lru.get(&path) {
+            let f = match lru.get_or_init()?.get(&path) {
                 Ok(f) => f,
                 Err(LruError::FileNotInCache) => {
                     trace!("DiskCache::get({}): FileNotInCache", key);
@@ -85,7 +117,7 @@ impl Storage for DiskCache {
         Box::new(self.pool.spawn_fn(move || {
             let start = Instant::now();
             let v = entry.finish()?;
-            lru.lock().unwrap().insert_bytes(key, &v)?;
+            lru.lock().unwrap().get_or_init()?.insert_bytes(key, &v)?;
             Ok(start.elapsed())
         }))
     }
@@ -95,9 +127,9 @@ impl Storage for DiskCache {
     }
 
     fn current_size(&self) -> SFuture<Option<u64>> {
-        f_ok(Some(self.lru.lock().unwrap().size()))
+        f_ok(self.lru.lock().unwrap().get().map(|l| l.size()))
     }
     fn max_size(&self) -> SFuture<Option<u64>> {
-        f_ok(Some(self.lru.lock().unwrap().capacity()))
+        f_ok(self.lru.lock().unwrap().get().map(|l| l.capacity()))
     }
 }
```

However, both this new patch and your patch suffer from the problem that the first compile is going to wait for the cache initialization, and if that takes more than 60s, will bail out.

The ideal solution would be to not have the cache initialization blocking reads and writes from the cache at all.